### ISSUE TITLE
Provide API for getting Material Icons by string key

### ISSUE THREAD

**Issue by alanrussian** [flutter/flutter#16189]
Right now, you have to reference Icons statically by calling the proper getter on the `Icons` class. It would be useful to be able to reference icons by their string key as well. 

One use case: sharing code between AngularDart and Flutter apps. Let's say you have a Dart function that returns the material icon that should be used for some argument. What do you return? For web, [you're required to specify the string key of the icon](https://pub.dartlang.org/documentation/angular_components/0.8.0/material_icon_material_icon/MaterialIconComponent-class.html) but Flutter doesn't let you do that.

**Comment by zoechi** [comment]
Isn't it easy enough to build your own map from icon name (string) to the enum value just for the icons you are using? This way you also can use names that match better with what you are using the icons for. 

```dart
const myIcons = <String, IconData> {
  'threesixty': Icons.threesixty,
  'threed_rotation': Icons.threed_rotation,
};

...

new Icon(myIcons['threed_rotation']);
```

**Comment by alanrussian** [comment]
Yeah that's easy but:

- I'm consuming an API for providing icon names that's provided by another
team. I can't assume they won't add icons over time.
- Even if I added all icons, more material icons are added over time.

In regards to your last sentence, there's a standard naming of these icons
that's consistent.

**Comment by zoechi** [comment]
Are there plans to do anything or can this issue be closed?

**Comment by alanrussian** [comment]
I understand the benefit of tree shaking, but there's a big benefit from the code sharing too. Would it be possible to make it so the icons aren't tree shaken if the Icons are accessed using non-const strings?

**Comment by jonahwilliams** [comment]
No, putting something in a map is using it accordingly to the tree shaker, that is unlikely to change.  There is nothing stopping you from making your own map of icons

**Comment by alanrussian** [comment]
Two people did thumbs up this issue though. Would it be possible to somehow make this an optional dependency so that the icons are tree shaken if you don't include the dependency?

**Comment by zoechi** [comment]
Such an optional dependency could be a published package by the community

**Comment by jonahwilliams** [comment]
I think a package is a perfectly reasonable approach to this

**Comment by 24601** [comment]
I have a VERY quick and dirty first pass at such a package. I do not claim that it's wonderfully designed & architected, but it does provide one-method access to grabbing IconData for Material & Font Awesome icons using (@brianegan's great package...). 

https://pub.dartlang.org/packages/icons_helper

**Comment by 24601** [comment]
FYI - noticed my above package I mentioned/created as a fix for this wasn't a good Dart 2 citizen. Updated.

**Comment by bitriddler** [comment]
This script will generate a file that exposes all available icons
https://gist.github.com/bitriddler/f09f9108be81f76475acb0f216e4f9bc

you will need to add this script `generate-available-icons.dart` in the `bin` folder and enable it in `pubspec.yaml`

```yaml
executables:
  generate-available-icons: 
```

Then you can run it with
```bash
FLUTTER_DOWNLOAD_PATH=[flutter-download-path] flutter pub run generate-available-icons
```

Not sure about tree shaking tho, I'm pretty new to dart/flutter but I would expect dart to eliminate unused icons...

**Comment by yudhi115** [comment]
"you will need to add this script generate-available-icons.dart in the bin folder and enable it in pubspec.yaml"
In my case the bin folder does not include "/packages/flutter/lib/src/material/icons.dart"
my "c:\src\flutter" has this path

**Comment by yudhi115** [comment]
great job!!!
